criu/sockets.c

```diff
@@ -425,48 +425,63 @@ static int restore_socket_filter(int sk, SkOptsEntry *soe)
 	return ret;
 }
 
-static struct socket_desc *sockets[SK_HASH_SIZE];
+static struct socket_table_entry *sockets[SK_HASH_SIZE];
 
-struct socket_desc *lookup_socket_ino(unsigned int ino, int family)
+struct socket_table_entry *lookup_socket_ino(unsigned int ino)
+{
+	struct socket_table_entry *e;
+
+	for (e = sockets[ino % SK_HASH_SIZE]; e; e = e->next) {
+		if (e->ino == ino)
+			return e;
+	}
```

Comment:
hlist_for_each_entry

Reply:
Thanks for the advice, I will try doing this, just one question.
Should the lookup table be initialized to NULL like [here](https://github.com/checkpoint-restore/criu/blob/328c1d540df827fa5abfbbc2056b7cdcb0648818/criu/files.c#L66)? I am asking because the previous implementation for socket_desc was not initializing the table and the tests passed successfully.
If it needs initialization, I guessing that could be done [here](https://github.com/checkpoint-restore/criu/blob/328c1d540df827fa5abfbbc2056b7cdcb0648818/criu/files.c#L1749), please correct me if I am wrong.

Reply:
@Snorch 
I faced a problem, I can't use `struct unix_sk_info` from `sockets.c` because it is only defined in `sk_unix.c`.
I am thinking of two solutions:
1. This is a workaround and I think it is ugly:
- Add a function to get the `hlist_head` by ino.
- Then `hlist_for_each_entry` to get the `struct unix_sk_info` entry in `sk_unix.c`.
2. Another solution is leaving the hashtable implementation for `struct socket_desc` as it was originally and creating a table in `sk_unix.c` only for `struct unix_sk_info`, this is not generalized but will work well.

btw I think `container_of` could be used `struct socket_table_entry` if we don't want it to be always be the first member.
maybe something like this:
```c
#define table_entry(ptr, type, member) container_of(ptr,type,member)
```
What is your opinion? Any suggestions?

Reply:
>Should the lookup table be initialized to NULL ...

Technically for file_desc_hash we can skip it because global variables are zero-initialized. But for me personally explicitly using INIT_HLIST_HEAD protects us from any internal changes in hlist (which will likely never happen I  know, but to be on the safe side).

>I am guessing that could be done here ...

Initializing hlists in prepare_files looks like a right place on restore, but you miss dump.

>I faced a problem, I can't use struct unix_sk_info from sockets.c because it is only defined in sk_unix.c. 1. ... 2. ...

I like (2) we don't want both unix_sk_info-s and socket_desc-s in one list anyway, why not separating lists for readability.

You should likely resurect find_unix_sk_by_ino, it will lookup through unix_sk_info-s and lookup_socket_ino will lookup through socket_desc-s. Something like:

```
static struct hlist_head sockets[SK_HASH_SIZE];

struct unix_sk_info {
	...
	struct hlist_node h;
};


static struct unix_sk_info *find_unix_sk_by_ino(int ino)
{
	struct unix_sk_info *ui;

	hlist_for_each_entry(ui, &sockets[ino % SK_HASH_SIZE], h) {
		if (ui->ue->ino == ino)
			...
	}
	...
}

struct socket_desc{
	...
	struct hlist_node h;
};

static struct socket_desc *lookup_socket_ino(int ino)
{
	struct socket_desc *d;

	hlist_for_each_entry(d, &sockets[ino % SK_HASH_SIZE], h) {
		if (d->ino == ino)
			...
	}
	...
}
```

Reply:
Done, But I think we should keep the PR focused on one thing at a time.
I only optimized ```find_unix_sk_by_ino()```, I think a separate PR should be made to refactor other hashtable implementations, like [this](https://github.com/checkpoint-restore/criu/blob/996a11d708fd5a5fc6166fe8cfaee80095e920a3/criu/sockets.c#L428) and [this](https://github.com/checkpoint-restore/criu/blob/996a11d708fd5a5fc6166fe8cfaee80095e920a3/criu/sk-unix.c#L118).